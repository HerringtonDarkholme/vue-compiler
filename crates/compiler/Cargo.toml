[package]
name = "compiler"
version = "0.1.0"
authors = ["Herrington Darkholme <2883231+HerringtonDarkholme@users.noreply.github.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
smallvec = "1.6.1"
bitflags = "1.3"
rustc-hash = "1.1.0"
json = "0.12.4"
anyhow = "1.0.44"
codespan-reporting = "0.11.1"

[dev-dependencies]
insta = "1.8.0"
serde = { version = "1.0", features = ["derive"] }
smallvec = {version="1.2.0", features = ["serde"]}
