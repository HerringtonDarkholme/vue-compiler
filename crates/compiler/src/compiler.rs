use super::{
    codegen::CodeGenerator,
    converter::Converter,
    error::ErrorHandler,
    parser::{ParseOption, Parser},
    tokenizer::{TokenizeOption, Tokenizer},
    transformer::Transformer,
};

pub struct CompileOption<E: ErrorHandler> {
    tokenization: TokenizeOption,
    parsing: ParseOption,
    error_handler: E,
}

pub trait TemplateCompiler<'a> {
    type IR;
    type Output;
    type Eh: ErrorHandler;
    type Conv: Converter<'a, IR = Self::IR>;
    type Trans: Transformer<IR = Self::IR>;
    type Gen: CodeGenerator<IR = Self::IR, Output = Self::Output>;

    fn get_tokenizer() -> Tokenizer;
    fn get_parser() -> Parser;
    fn get_converter() -> Self::Conv;
    fn get_transformer() -> Self::Trans;
    fn get_code_generator() -> Self::Gen;
    fn get_error_handler() -> Self::Eh;

    fn compile(source: &'a str) -> Self::Output {
        let tokenizer = Self::get_tokenizer();
        let parser = Self::get_parser();
        let eh = Self::get_error_handler();
        let tokens = tokenizer.scan(source, eh);
        let eh = Self::get_error_handler();
        let ast = parser.parse(tokens, eh);
        let mut ir = Self::get_converter().convert_ir(ast);
        Self::get_transformer().transform(&mut ir);
        Self::get_code_generator().generate(ir)
    }
}

pub fn base_compile<'a, C: TemplateCompiler<'a>>(source: &'a str) -> C::Output {
    C::compile(source)
}
