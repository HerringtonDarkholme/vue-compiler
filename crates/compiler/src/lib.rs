#![allow(dead_code, unused_variables)]
//! See README.md

// TODO: reorg pub
pub mod codegen;
pub mod converter;
pub mod error;
pub mod flags;
pub mod parser;
pub mod tokenizer;
pub mod transformer;
pub mod util;

use std::ops::Range;

pub use codegen::CodeGenerator;
pub use converter::Converter;
use error::{CompilationError, ErrorHandler};
use parser::{ParseOption, Parser};
use tokenizer::{TokenizeOption, Tokenizer};
pub use transformer::Transformer;

// use plain &str here for now
// may change to tendril
pub type Name<'a> = &'a str;

#[derive(PartialEq, Eq, Clone)]
pub struct Position {
    /// the 0-indexed offset in the source str modulo newline
    pub offset: usize,
    /// the line number in the source code
    pub line: usize,
    /// the column number in the source code
    pub column: usize,
}

impl Default for Position {
    fn default() -> Self {
        Self {
            offset: 0,
            line: 1,
            column: 1,
        }
    }
}

#[derive(Default, PartialEq, Eq, Clone)]
pub struct SourceLocation {
    pub start: Position,
    pub end: Position,
}

impl Into<Range<usize>> for SourceLocation {
    fn into(self) -> Range<usize> {
        self.start.offset..self.end.offset
    }
}

/// namespace for HTML/SVG/MathML tag
#[non_exhaustive]
#[derive(Eq, PartialEq)]
pub enum Namespace {
    Html,
    Svg,
    MathMl,
    UserDefined(&'static str),
}

pub trait TemplateCompiler {}

pub struct CompileOption<E: ErrorHandler> {
    tokenization: TokenizeOption,
    parsing: ParseOption,
    error_handler: E,
}

pub fn base_compile<'a, IR, O, E, Conv, Trans, Gen>(
    source: &'a str,
    opt: CompileOption<E>,
    conv: Conv,
    trans: Trans,
    mut gen: Gen,
) -> Result<O, CompilationError>
where
    E: ErrorHandler + Clone,
    Conv: Converter<'a, IR = IR>,
    Trans: Transformer<IR = IR>,
    Gen: CodeGenerator<IR = IR, Output = O>,
{
    let eh = opt.error_handler;
    let tokenizer = Tokenizer::new(opt.tokenization);
    let parser = Parser::new(opt.parsing);
    let tokens = tokenizer.scan(source, eh.clone());
    let ast = parser.parse(tokens, eh);
    let mut ir = conv.convert_ir(ast);
    trans.transform(&mut ir);
    Ok(gen.generate(ir))
}
