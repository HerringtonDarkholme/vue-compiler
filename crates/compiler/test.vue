<template >
    <div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div v-if=></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
    </div>
</template>
