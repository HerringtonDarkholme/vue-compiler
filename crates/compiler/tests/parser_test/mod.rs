mod dir;
use std::fs::read_to_string;

use super::common::{serialize_yaml, TestErrorHandler};
use super::tokenizer_test::base_scan;
use compiler::parser::{self as p, ParseOption, Parser};
use compiler::tokenizer::TextMode;
use insta::assert_snapshot;

fn test_ast(case: &str) {
    let name = insta::_macro_support::AutoName;
    let root = base_parse(case);
    let val = serialize_yaml(root);
    assert_snapshot!(name, val, case);
}

#[test]
fn test_base_parse() -> std::io::Result<()> {
    let cases = ["<p/>", "<p></p>", "<p>123</p>"];
    for case in cases {
        test_ast(case);
    }
    Ok(())
}

#[test]
fn test_base_file_parse() -> std::io::Result<()> {
    let file_list = ["basic.vue"];
    for file_name in file_list {
        let file = read_to_string(format!("tests/test_files/{}", file_name))?;
        test_ast(&file);
    }
    Ok(())
}

pub fn base_parse(s: &str) -> p::AstRoot {
    let tokens = base_scan(s);
    let parser = Parser::new(ParseOption {
        get_text_mode: |s| match s {
            "script" => TextMode::RawText,
            "textarea" => TextMode::RcData,
            _ => TextMode::Data,
        },
        is_native_element: |s| s != "comp",
        ..ParseOption::default()
    });
    let eh = TestErrorHandler;
    parser.parse(tokens, eh)
}
