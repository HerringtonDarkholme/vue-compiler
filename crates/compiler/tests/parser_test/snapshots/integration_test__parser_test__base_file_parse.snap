---
source: crates/compiler/tests/parser_test/mod.rs
expression: "<template>\n  <div v-for=\"item in result\">\n    <p>test {{item}}</p>\n    <input v-model=\"inputValue\"/>\n  </div>\n</template>\n"

---
children:
  - Element:
      tag_name: template
      tag_type: Plain
      namespace: Html
      properties: []
      children:
        - Element:
            tag_name: div
            tag_type: Plain
            namespace: Html
            properties:
              - Dir:
                  name: for
                  argument: ~
                  modifiers: []
                  expression:
                    content: item in result
                    location:
                      start: "Pos: 24, Ln: 2, Col: 14"
                      end: "Pos: 40, Ln: 2, Col: 30"
                  head_loc:
                    start: "Pos: 18, Ln: 2, Col: 8"
                    end: "Pos: 23, Ln: 2, Col: 13"
                  location:
                    start: "Pos: 18, Ln: 2, Col: 8"
                    end: "Pos: 40, Ln: 2, Col: 30"
            children:
              - Element:
                  tag_name: p
                  tag_type: Plain
                  namespace: Html
                  properties: []
                  children:
                    - Text:
                        text: "test "
                        location:
                          start: "Pos: 49, Ln: 3, Col: 8"
                          end: "Pos: 54, Ln: 3, Col: 13"
                    - Interpolation:
                        source: item
                        location:
                          start: "Pos: 54, Ln: 3, Col: 13"
                          end: "Pos: 62, Ln: 3, Col: 21"
                  location:
                    start: "Pos: 46, Ln: 3, Col: 5"
                    end: "Pos: 66, Ln: 3, Col: 25"
              - Element:
                  tag_name: input
                  tag_type: Plain
                  namespace: Html
                  properties:
                    - Dir:
                        name: model
                        argument: ~
                        modifiers: []
                        expression:
                          content: inputValue
                          location:
                            start: "Pos: 86, Ln: 4, Col: 20"
                            end: "Pos: 98, Ln: 4, Col: 32"
                        head_loc:
                          start: "Pos: 78, Ln: 4, Col: 12"
                          end: "Pos: 85, Ln: 4, Col: 19"
                        location:
                          start: "Pos: 78, Ln: 4, Col: 12"
                          end: "Pos: 98, Ln: 4, Col: 32"
                  children: []
                  location:
                    start: "Pos: 71, Ln: 4, Col: 5"
                    end: "Pos: 100, Ln: 4, Col: 34"
            location:
              start: "Pos: 13, Ln: 2, Col: 3"
              end: "Pos: 109, Ln: 5, Col: 9"
      location:
        start: "Pos: 0, Ln: 1, Col: 1"
        end: "Pos: 121, Ln: 6, Col: 12"
location:
  start: "Pos: 0, Ln: 1, Col: 1"
  end: "Pos: 122, Ln: 7, Col: 1"

